### PR TITLE
[WIP] TisTos

### PR DESCRIPTION
**Pull request by roelaaij**
Added empty file for TisTos lesson, working on it.

### CODE REVIEW

**Comment by betatim** [comment]
Unsure what you are planning, so this comment might become irrelevant.

I'd like the style of the workshop to be something like 10min (max) of talking by the instructor and then the students go off and do something, then after they did something we come back to some more talking but quickly they do something again. The idea being that you don't learn anything unless you do something.

Long chunk of text is good to have to explain stuff, we should just think of either making another lesson that has lots of examples (and we keep the long text as "advanced, study at your leisure" material) or we add lots of examples to this. (Yes, it is easier to say than do for some cases)

**Comment by betatim** [comment]
@roelaaij is a :racehorse: wanted to comment on how to fix the build but already fixed!

**Comment by roelaaij** [comment]
I tried to keep everything very concise, but didn't want to leave anything as "just do this and then it will work", so the lesson got a bit on the long side.

**Comment by betatim** [comment]
Should we split it up? I was going to ask what we do about the other three HLT themed lessons. Might be worht splitting things a bit to give a natural place for a short break and synchronise everyone.

**Comment by roelaaij** [comment]
I can split it before/after having a look at the LHCbIDs of some stored candidates and call the first one the HLT intro and the other one TisTos DIY, or something like that.

File: 18-hlt-intro.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+> * Learn a bit about the LHCb trigger
+> * Learn how the results of the HLT are stored in the raw event.
+> * Use an interactive python session to explore the HLT results.
+
+The LHCb trigger reduces the input rate event rate of approximately 12 MHz
+to 12.5 kHz of events that are written to offline tape storage. The rate
+reduction is achieved in three steps: L0, HLT1 and HLT2. L0 is implemented in
+custom FPGAs, has a fixed latency of 4 μs and a maximum output rate of 1 MHz.
+
+HLT1 and HLT2 are implemented as a software applications that run on the online
```
Comment (betatim) [inline 18-hlt-intro.md]: "as a" -> "as"

File: 18-hlt-intro.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+Once HLT1 or HLT2 has accepted an event, the candidates accepted by all trigger
+lines are saved to the raw event in a stripped-down form. One of the things that
+is saved are all the LHCbIDs of the final state particles of a decay tree.
+LHCbIDs are unique identifiers for detector hits.
+
+We will now have a look at the LHCbIDs of some of the stored candidates. We will
+use the script we [created earlier](05-interactive-dst.html) as a starting
+point, and the file you [downloaded before](05-files-from-grid.html). Fire up
+your favourite editor, open the script and save a copy to work on as
+"tistos.py". There are a few things in the script that we don't need and can be
```
Comment (betatim) [inline 18-hlt-intro.md]: Change the " to `

File: 18-hlt-intro.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+removed, such as the print_decay and decay finder tools.
+
+Like the stripping, the decisions of the HLT are saved in so-called
+DecReports. You can find them in "Hlt/DecReports" and "Hlt2/DecReports", have a
```
Comment (betatim) [inline 18-hlt-intro.md]: DecReports -> `DecReports` and " -> `

File: 18-hlt-intro.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+look at what they contain.
+
+~~~ {.python}
+evt['Hlt1/DecReports']
+evt['Hlt2/DecReports']
+~~~
+
+An important difference with the stripping is that for the HLT, all the
+decisions are present, even if they are false. Copy the advance function to
```
Comment (betatim) [inline 18-hlt-intro.md]: advance -> `advance`

File: 18-hlt-intro.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+advance_hlt, make it work on HLT decisions and make sure it really checks the
```
Comment (betatim) [inline 18-hlt-intro.md]: advance_hlt -> `advance_hlt`

File: 18-hlt-intro.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+decision. This can be done using the "decision" member function of a
```
Comment (betatim) [inline 18-hlt-intro.md]: "decision" -> `decision`

File: 18-hlt-intro.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+DecReport. Note that all names in the reports end with "Decision".
```
Comment (betatim) [inline 18-hlt-intro.md]: `DecReport`

**Comment by betatim** [comment]
I'll hold off from making line comments on "Blah" -> `Blah` transforms. Backticks are the markdown formatting for code, so all things code-y should use them

File: 18-hlt-intro.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+In addition to the LHCbIDs, some numbers are also stored, such as the momentum
+of the track. These are stored in the numerical info dictionary that can be
+retrieved using:
+
+~~~ {.python}
+report.substructure().substructure()[0].numericalInfo()
+~~~
+
+Bonus exercise: Make a plot of the transverse momentum distribution of
```
Comment (betatim) [inline 18-hlt-intro.md]: To get a nice HTML box with colour:

```
> ## Plot transverse momentum distribution {.challenge}
>
> Make a plot of the pT dist ...
```

Comment (betatim) [inline 18-hlt-intro.md]: It doesn't render right on GH, works for our generated HTML though. Trust me(tm)

File: 18-hlt-intro.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+candidates accepted by the Hlt1TrackAllL0 selection.
+
+The LHCbIDs of the (in this case) track can be retrieved using:
+
+~~~ {.python}
+report.substructure().substructure()[0].lhcbIDs()
+~~~
+
+Excercise: The LHCbIDs can tell you which sub-detector they belong too. Have a
```
Comment (betatim) [inline 18-hlt-intro.md]:
```
> ## Subdetector LHCb IDs {.challenge}
>
> The LHCbIDs can tell you ...
```

File: 18-tistos-diy.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+To have a look at how this works, we'll use candidates from the
+"D2hhPromptDst2D2RS" selection, which can be retrieved thusly:
+
+~~~ {.python}
+candidates = evt['AllStreams/Phys/D2hhPromptDst2D2RSLine/Particles']
+candidates.size()
+~~~
+
+I could be that there are more than one candidates, which are unlikely to all be
```
Comment (betatim) [inline 18-tistos-diy.md]: I could be -> It could be

File: 18-tistos-diy.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+result = hlt1TisTosTool.tisTosTobTrigger()
+~~~
+
+The (Tos) trigger efficiency of a trigger selection can be calculated as:
+
+eps\_{Tos} = N\_{Tis AND Tos} / N\_{Tis}
+
+Loop over the events in the DST and calculate the efficiency of
+Hlt1TrackAllL0. You can add some more Hlt1 selecitons when checking for Tis,
+which ones would make sense?
```
Comment (betatim) [inline 18-tistos-diy.md:112]: (Why do you not use "all other physics" triggers when looking for Tis? This is something I never understood. Maybe I should ask during the workshop ;) )

**Comment by betatim** [comment]
Good work, I like it. I think it will make it much more comprehendible what all this TIS and TOS stuff is about.

Most of the comments are about formatting. Tedious but necessary (if we don't fix it now we will never fix it).

**Comment by roelaaij** [comment]
I've adressed the comments, could this PR be merged? Let me know something is still needed.

**Comment by betatim** [comment]
LGTM.

(Note to future: we will probably move this to `second-analysis-steps` once that exists)

**Comment by betatim** [comment]
Thanks a lot for this!! Especially with #13TeV happening at the same time.

Time for :sunglasses:

**Pull request merged**